piet-gpu/shader/kernel4.comp

```diff
@@ -35,29 +35,7 @@ layout(rgba8, set = 0, binding = 3) uniform readonly image2D images[1];
 #include "ptcl.h"
 #include "tile.h"
 
-#define BLEND_STACK_SIZE 4
-
-// Layout of a clip scratch frame:
-// Each frame is WIDTH * HEIGHT 32-bit words, then a link reference.
-
-// Link offset and frame size in 32-bit words.
-#define CLIP_LINK_OFFSET (TILE_WIDTH_PX * TILE_HEIGHT_PX)
-#define CLIP_BUF_SIZE (CLIP_LINK_OFFSET + 1)
-
-shared MallocResult sh_clip_alloc;
-
-// Allocate a scratch buffer for clipping.
-MallocResult alloc_clip_buf(uint link) {
-    if (gl_LocalInvocationID.x == 0 && gl_LocalInvocationID.y == 0) {
-        MallocResult m = malloc(CLIP_BUF_SIZE * 4);
-        if (!m.failed) {
-            write_mem(m.alloc, (m.alloc.offset >> 2) + CLIP_LINK_OFFSET, link);
-        }
-        sh_clip_alloc = m;
-    }
-    barrier();
-    return sh_clip_alloc;
-}
+#define MAX_BLEND_STACK 256
```

Comment:
The old code didn't have a limit on the stack size, however large. It may be ok, but there are few (no?) other hard limits in piet-gpu.

We can comoue the worst case stack depth during encode; can that information be used here somehow? Perhaps a specialization constant, but I'd rather like to avoid dynamic shader source to make it easier to do CPU fallback.

Reply:
It's a rather unfortunate limitation of this approach, and no, I don't think we can change the allocation size of the scratch buffer without dynamic shader in a portable way. Even then, 256 is a quite conservative value, so I hope it will never be required to be raised.